File: java/LinkedList/linkedList.java
package linkedList;
import java.util.Scanner;

class Node {
    int data;
    Node next;
    Node pre;

    Node(int d) {
        this.data = d;
        this.next = null;
        this.pre = null;
    }
}

class Linking {
    Node head;
    public void insert(int data, int index) {
        if(index == 0) {
            addInTheBegining(data);
            return;
        }
        Node nNode = new Node(data);
        Node temp = this.head;
        int count = 1;
        while(count != index) {
            temp = temp.next;
            count++;
        }
        nNode.next = temp.next;
        temp.next = nNode;
    }
    public void addInTheBegining(int item) {
        Node newNode = new Node(item);
        if(this.head == null) {
            this.head = newNode;
            return;
        }
        newNode.next = this.head;
        this.head = newNode;
        
    }
    public void addInTheEnd(int item) {
        Node newNode = new Node(item);
        if(this.head == null) {
            this.head = newNode;
            return;
        }
        Node current = this.head;
        while (current.next!=null) {
            current = current.next;
        }
        current.next = newNode;
        
    }

    public void  delete (int key) {
        Node temp = this.head;
        while(temp.next != null){
            if (temp.next.data == key) {
                temp.next = temp.next.next;
                System.out.println("The element " + key + " deleted");
                break;
            }
            temp = temp.next;
        }
        System.out.println("Not present in this list");
        
    }
    public void display() {
        Node current = this.head;
        while (current!=null) {
            System.out.print(current.data + "->");
            current = current.next;
        }
    }

    
}

public class linkedList {    
    public void nu() {
        
    }
    public static void main(String args[]) {
        //main block
        Scanner In = new Scanner(System.in);
        Linking l = new Linking();
        
        while (true) { 
            System.out.println("\nEnter Operation:");
            System.out.println("1 for insert with index");
            System.out.println("2 for adding in the begining");
            System.out.println("3 for adding in the end");
            System.out.println("4 for delete with data");
            System.out.println("5 for diplay");
            System.out.println("6 for exit");

            int op = In.nextInt();
            if (op == 1) {
                System.out.println("Enter the index:");
                int i = In.nextInt();
                System.out.println("Enter the data:");
                int d = In.nextInt();
                l.insert(d, i);
            } else if (op == 2) {
                System.out.println("Enter the data:");
                int d = In.nextInt();
                l.addInTheBegining(d);
            } else if (op == 3) {
                System.out.println("Enter the data:");
                int d = In.nextInt();
                l.addInTheEnd(d);
            } else if (op == 4) {
                System.out.println("Enter the data which you want to remove :");
                int d = In.nextInt();
                l.delete(d);
            } else if (op==5){
                l.display();
            } else if (op == 6) {
                System.exit(0);
            } else {
                System.out.println("Enter a valid Number!");
            }
            
            In.close();
        }
        
    }
}

File: cpp/LinkedList/LinkedList.cpp
#include <iostream>
using namespace std;

class Node {
public:
    int val;
    Node* next;
    Node(int val) {
        this->val = val;
        this->next = nullptr;
    }
};

class LinkedList {
public:
    Node* head;
    LinkedList() {
        this->head = nullptr;
    }

    void append(int val) {
        Node* newNode = new Node(val);
        if (this->head == nullptr) {
            this->head = newNode;
            return;
        }
        Node* temp = this->head;
        while (temp->next != nullptr) {
            temp = temp->next;
        }
        temp->next = newNode;
    }

    void prepend(int val) {
        Node* newNode = new Node(val);
        if (this->head == nullptr) {
            this->head = newNode;
            return;
        }
        newNode->next = this->head;
        this->head = newNode;
    }

    void deleteNode(int val) {
        if (this->head == nullptr) {
            return;
        }

        if(this->head->val == val) {
            Node* temp = this->head;
            this->head = this->head->next;
            delete temp;
            return;
        }

        Node* temp = this->head;

        while(temp->next != nullptr && temp->next->val != val) {
            temp = temp->next;
        }

        if(temp->next == nullptr) {
            return;
        }

        Node* nodeToDelete = temp->next;
        temp->next = temp->next->next;
        delete nodeToDelete;
    }

    void display() {
        Node* temp = this->head;
        while (temp != nullptr) {
            cout << temp->val << " ";
            temp = temp->next;
        }
        cout << endl;
    }

    bool empty() {
        return this->head == nullptr;
    }

    int size() {
        int count = 0;
        Node* temp = this->head;
        while (temp != nullptr) {
            count++;
            temp = temp->next;
        }
        return count;
    }

    int search(int val) {
        Node* temp = this->head;
        int index = 0;
        while (temp != nullptr) {
            if (temp->val == val) {
                return index;
            }
            temp = temp->next;
            index++;
        }
        return -1; // Value not found
    }

};

int main () {
    LinkedList list;
    cout << "Is list empty? " << list.empty() << endl; // Output: Is list empty? 1
    list.append(10);
    list.append(20);
    list.append(30);
    list.display(); // Output: 10 20 30
    list.prepend(5);
    list.display(); // Output: 5 10 20 30
    list.deleteNode(20);
    list.display(); // Output: 5 10 30
    cout << "Size: " << list.size() << endl; // Output: Size: 3
    cout << "Search for 10: " << list.search(10) << endl; // Output: Search for 10: 1
    cout << "Search for 40: " << list.search(40) << endl; // Output: Search for 40: -1
    return 0;
}